compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts

```diff
@@ -0,0 +1,370 @@
+   *    shared.set(id, node);
+   *  } else {
+   *    emit(env, locals, shared, nextInstructions, id);
+   *  }
+   *}
+   *if (block.kind !== "block" && block.kind !== "catch") {
+   *  const extra = nextInstructions.splice(index);
+   *  nextInstructions.splice(0, 0, ...extra);
+   *}
+   */
+  block.instructions = nextInstructions;
+  DEBUG && console.log();
+}
+
+function getDepth(env: Environment, nodes: Nodes, id: IdentifierId): number {
+  const node = nodes.get(id)!;
+  if (node == null) {
+    return 0;
+  }
+  if (node.depth != null) {
+    return node.depth;
+  }
+  node.depth = 0; // in case of cycles
+  let depth =
+    node.instruction != null && mayAllocate(env, node.instruction) ? 1 : 0;
+  for (const dep of node.dependencies) {
+    depth += getDepth(env, nodes, dep);
+  }
+  node.depth = depth;
+  return depth;
+}
+
+function print(
+  env: Environment,
+  locals: Nodes,
+  shared: Nodes,
+  seen: Set<IdentifierId>,
+  id: IdentifierId,
+  depth: number = 0
+): void {
+  if (seen.has(id)) {
+    console.log(`${"|   ".repeat(depth)}$${id} <skipped>`);
+    return;
+  }
+  seen.add(id);
+  const node = locals.get(id) ?? shared.get(id);
+  if (node == null) {
+    return;
+  }
+  const deps = [...node.dependencies];
+  deps.sort((a, b) => {
+    const aDepth = getDepth(env, locals, a);
+    const bDepth = getDepth(env, locals, b);
+    return bDepth - aDepth;
+  });
+  for (const dep of deps) {
+    print(env, locals, shared, seen, dep, depth + 1);
+  }
+  console.log(
+    `${"|   ".repeat(depth)}$${id} ${printNode(node)} deps=[${deps.map((x) => `$${x}`).join(", ")}]`
+  );
+}
+
+function printNode(node: Node): string {
+  const { instruction } = node;
+  if (instruction === null) {
+    return "<lvalue-only>";
+  }
+  switch (instruction.value.kind) {
+    case "FunctionExpression":
+    case "ObjectMethod": {
+      return `[${instruction.id}] ${instruction.value.kind}`;
+    }
+    default: {
+      return printInstruction(instruction);
+    }
+  }
+}
+
+function emit(
+  env: Environment,
+  locals: Nodes,
+  shared: Nodes,
+  instructions: Array<Instruction>,
+  id: IdentifierId
+): void {
+  const node = locals.get(id) ?? shared.get(id);
+  if (node == null) {
+    return;
+  }
+  locals.delete(id);
+  shared.delete(id);
+  const deps = [...node.dependencies];
+  deps.sort((a, b) => {
+    const aDepth = getDepth(env, locals, a);
+    const bDepth = getDepth(env, locals, b);
+    return bDepth - aDepth;
+  });
+  for (const dep of deps) {
+    emit(env, locals, shared, instructions, dep);
+  }
+  if (node.instruction !== null) {
+    instructions.push(node.instruction);
+  }
+}
+
+enum Reorderability {
+  Reorderable,
+  Nonreorderable,
+}
+function getReoderability(instr: Instruction): Reorderability {
+  switch (instr.value.kind) {
+    case "JsxExpression":
+    case "JsxFragment":
+    case "JSXText":
+    case "LoadGlobal":
+    case "Primitive":
+    case "TemplateLiteral": {
+      return Reorderability.Reorderable;
```

Comment:
BinaryExpression?

Reply:
Oh good call. UnaryExpression too

Reply:
Was uncertain about this but makes sense--opened #29880 to resolve my uncertainty

Reply:
nice! Yeah, we had already split out PropertyDelete but i thought that `throw` as an expression wasn't part of the spec yet? but in any case, good to have a todo for it in advance!